presto-native-execution/presto_cpp/main/runtime-metrics/RuntimeMetricsReporter.cpp

```
@@ -0,0 +1,70 @@
+#include "presto_cpp/main/common/Configs.h"
+#include "presto_cpp/main/runtime-metrics/prometheus-reporter/PrometheusReporterUtil.h"
+#include "velox/velox/common/base/StatsReporter.h"
+
+namespace facebook::presto {
+namespace {
+
+std::unordered_map<
+    std::string,
+    std::unique_ptr<facebook::velox::BaseStatsReporter>>&
+metricsReporters() {
+  static std::unordered_map<
+      std::string,
+      std::unique_ptr<facebook::velox::BaseStatsReporter>>
+      metricsReporters;
+  return metricsReporters;
+}
+} // namespace
+
+void registerRuntimeMetricsHttpEndpoint(
+    const std::unique_ptr<http::HttpServer>& httpServer) {
+  // Add compile flags.
+  httpServer->registerGet(
+      "/v1/info/metrics/",
+      [](proxygen::HTTPMessage* /*message*/,
+         const std::vector<std::unique_ptr<folly::IOBuf>>& /*body*/,
+         proxygen::ResponseHandler* downstream) {
+        auto systemConfig = SystemConfig::instance();
+        auto reporterFormat = systemConfig->runtimeMetricsReporterFormat();
+        if (reporterFormat == kPrometheusReporterName) {
+          http::sendOkResponse(downstream, prometheus::getSerializedMetrics());
+        }
+      });
+}
+
+void registerRuntimeMetricsReporters(
+    const std::unique_ptr<http::HttpServer>& httpServer) {
+  facebook::velox::BaseStatsReporter::registered = true;
+  metricsReporters().insert(
+      {kPrometheusReporterName.data(),
+       std::move(prometheus::createPrometheusReporter())});
+  metricsReporters().insert(
+      {kDummyReporterName.data(),
+       std::make_unique<facebook::velox::DummyStatsReporter>()});
+  registerRuntimeMetricsHttpEndpoint(httpServer);
+}
+
+// Initialize singleton for the reporter
+folly::Singleton<facebook::velox::BaseStatsReporter> reporter(
+    []() -> velox::BaseStatsReporter* {
+      auto systemConfig = SystemConfig::instance();
+      auto reporterName = systemConfig->runtimeMetricsReporterFormat();
+      return metricsReporters()[reporterName].get();
```

Should throw an error here or inside if Prometheus is requested but is not enabled via CMake.

---

The error handling can be inside `getSerializedMetrics()` as well

---

Yes, that is how the current error handling is. We just return an error message stating the metrics reporting is not registered.

```
  return kPrometheusReporterNotEnabled.data();
```